optuna/multi_objective/samplers/_nsga2.py

```diff
@@ -0,0 +1,254 @@
+
+        return parent_generation, parent_population
+
+    def _select_parent(
+        self,
+        study: "multi_objective.study.MultiObjectiveStudy",
+        population: List["multi_objective.trial.FrozenMultiObjectiveTrial"],
+    ) -> "multi_objective.trial.FrozenMultiObjectiveTrial":
+        candidate0 = self._rng.choice(population)
+        candidate1 = self._rng.choice(population)
+
+        # TODO(ohta): Consider crowding distance.
+        if candidate0._dominates(candidate1, study.directions):
+            return candidate0
+        else:
+            return candidate1
+
+
+def _fast_non_dominated_sort(
+    population: List["multi_objective.trial.FrozenMultiObjectiveTrial"],
+    directions: List[optuna.study.StudyDirection],
+) -> List[List["multi_objective.trial.FrozenMultiObjectiveTrial"]]:
+    dominated_count = defaultdict(int)  # type: DefaultDict[int, int]
+    dominates_list = defaultdict(list)
+
+    for p, q in itertools.combinations(population, 2):
+        if p._dominates(q, directions):
+            dominates_list[p.number].append(q.number)
+            dominated_count[q.number] += 1
+        elif q._dominates(p, directions):
+            dominates_list[q.number].append(p.number)
+            dominated_count[p.number] += 1
+
+    population_per_rank = []
+    while len(population) > 0:
+        non_dominated_population = []
+        i = 0
+        while i < len(population):
+            if dominated_count[population[i].number] == 0:
+                individual = population[i]
+                if i == len(population) - 1:
+                    population.pop()
+                else:
+                    population[i] = population.pop()
+                non_dominated_population.append(individual)
+            else:
+                i += 1
+
+        for x in non_dominated_population:
+            for y in dominates_list[x.number]:
+                dominated_count[y] -= 1
+
+        assert non_dominated_population != []
+        population_per_rank.append(non_dominated_population)
+
+    return population_per_rank
+
+
+def _crowding_distance_sort(
+    population: List["multi_objective.trial.FrozenMultiObjectiveTrial"],
+) -> None:
+    distances = defaultdict(float)
+    for i in range(len(population[0].values)):
+        population.sort(key=lambda x: x.values[i])
+
+        v_min = population[0].values[i]
+        v_max = population[-1].values[i]
+        assert v_min is not None
+        assert v_max is not None
+
+        width = v_max - v_min
+        if width == 0:
+            continue
+
+        distances[population[0].number] = float("inf")
+        distances[population[-1].number] = float("inf")
+
+        for j in range(1, len(population) - 1):
+            v_high = population[j + 1].values[i]
+            v_low = population[j - 1].values[i]
+            assert v_high is not None
+            assert v_low is not None
+
+            distances[population[j].number] += (v_high - v_low) / width
```

Comment:
Please rename `distances` to something like `l1_distances` or `manhattan_distances`, or maybe `normalized_l1_distances`.

Reply:
Thank you for the suggestion. I chose  `manhatten_distances` because it is the most frequently used word in the GitHub repositories among the three names.
Commit: https://github.com/optuna/optuna/pull/1163/commits/cc48c99147b9672e32e98251c206aa0b7664634c